homeassistant/components/sensor/utility_meter.py

```diff
@@ -0,0 +1,274 @@
+from homeassistant.components.sensor import (DOMAIN, PLATFORM_SCHEMA)
+from homeassistant.const import (
+    CONF_NAME, ATTR_UNIT_OF_MEASUREMENT, ATTR_ENTITY_ID,
+    EVENT_HOMEASSISTANT_START, STATE_UNKNOWN, STATE_UNAVAILABLE)
+from homeassistant.core import callback
+from homeassistant.helpers.event import (
+    async_track_state_change, async_track_time_change)
+from homeassistant.helpers.dispatcher import (
+    dispatcher_send, async_dispatcher_connect)
+from homeassistant.helpers.restore_state import RestoreEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_SOURCE_ID = 'source'
+ATTR_STATUS = 'status'
+ATTR_PERIOD = 'meter_period'
+ATTR_LAST_PERIOD = 'last_period'
+ATTR_LAST_RESET = 'last_reset'
+
+SIGNAL_START_PAUSE_METER = 'utility_meter_start_pause'
+SIGNAL_RESET_METER = 'utility_meter_reset'
+
+SERVICE_START_PAUSE = 'utility_meter_start_pause'
+SERVICE_RESET = 'utility_meter_reset'
+
+HOURLY = 'hourly'
+DAILY = 'daily'
+WEEKLY = 'weekly'
+MONTHLY = 'monthly'
+YEARLY = 'yearly'
+
+METER_TYPES = [HOURLY, DAILY, WEEKLY, MONTHLY, YEARLY]
+
+SERVICE_METER_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
+})
+
+CONF_SOURCE_SENSOR = 'source'
+CONF_METER_TYPE = 'cycle'
+CONF_METER_OFFSET = 'offset'
+CONF_PAUSED = 'paused'
+
+ICON = 'mdi:counter'
+
+PRECISION = 3
+PAUSED = "paused"
+COLLECTING = "collecting"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SOURCE_SENSOR): cv.entity_id,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_METER_TYPE): vol.In(METER_TYPES),
+    vol.Optional(CONF_METER_OFFSET, default=0): cv.positive_int,
+    vol.Optional(CONF_PAUSED, default=False): cv.boolean,
+})
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up the utility meter sensor."""
+    def run_setup(event):
+        """Delay the setup until Home Assistant is fully initialized."""
+        conf = discovery_info if discovery_info else config
+
+        meter = UtilityMeterSensor(hass,
+                                   conf[CONF_SOURCE_SENSOR],
+                                   conf.get(CONF_NAME),
+                                   conf.get(CONF_METER_TYPE),
+                                   conf.get(CONF_METER_OFFSET),
+                                   conf.get(CONF_PAUSED))
+
+        async_add_entities([meter])
+
+        @callback
+        def async_start_pause_meter(service):
+            """Process service start_pause meter."""
+            for entity in service.data[ATTR_ENTITY_ID]:
+                dispatcher_send(hass, SIGNAL_START_PAUSE_METER, entity)
+
+        @callback
+        def async_reset_meter(service):
+            """Process service reset meter."""
+            for entity in service.data.get(ATTR_ENTITY_ID):
+                dispatcher_send(hass, SIGNAL_RESET_METER, entity)
+
+        hass.services.async_register(DOMAIN, SERVICE_START_PAUSE,
```

Comment:
Sensors should be read only and not register services. If we want to change that it requires an architecture issue.

Reply:
I think there is a misunderstanding. The sensor tracks other sensors (just like the template sensor), what the service provides is a mean for automations to enable/disable the said tracking. 

This sensor is an implementation of existing smart meters (at least in Portugal our smart meters have multiple counters that start/pause based on time). What I did here with the services is to make these counters start/stop based on an external automation instead of creating a complex configuration that would limit the usefullness of this sensor beyond 1 or 2 utility companies in Portugal.

Reply:
I understand what this is. But it is still not in line with current architecture.

Reply:
Then it is me who doesn't understand 😞 

I will open the Issue in the architecture repo, but first need to understand what change am I asking for. Is it: 
- Adding services to sensors ?

Reply:
Yes! So far there are no services for the sensor component. We have said no many times before.

Reply:
I can't find any valid implementations of the examples above.

- The mopar sensor service sends a command to change the state of the car. Clearly against our architecture. That should be done from a switch platform or from a new mopar component.
- dovado service that sends an sms should be done from a new dovado component.
- The other four platforms' services are now redundant, after we implemented the general entity update service. Those services should be removed.